util/repr-bytes/src/lib.rs

```diff
@@ -0,0 +1,492 @@
+    ($mytype:ty) => {
+        impl<'a> TryFrom<&'a [u8]> for $mytype {
+            type Error = <Self as $crate::ReprBytes>::Error;
+            fn try_from(src: &'a [u8]) -> Result<Self, Self::Error> {
+                if src.len() != <Self as $crate::ReprBytes>::size() {
+                    return Err(Self::Error::from($crate::LengthMismatch {
+                        expected: <Self as $crate::ReprBytes>::size(),
+                        found: src.len(),
+                    }));
+                }
+                <$mytype as $crate::ReprBytes>::from_bytes(<$crate::GenericArray<
+                    u8,
+                    <Self as $crate::ReprBytes>::Size,
+                >>::from_slice(src))
+            }
+        }
+    };
+}
+
+/// Derive prost::Message from a ReprBytes implementation
+/// The corresponding protobuf has exactly one member, of type `bytes`.
+#[macro_export]
+macro_rules! derive_prost_message_from_repr_bytes {
+    ($mytype:ty) => {
+        impl ::prost::Message for $mytype {
+            #[inline]
+            fn encode_raw<B>(&self, buf: &mut B)
+            where
+                B: ::prost::bytes::BufMut,
+            {
+                use ::prost::encoding::*;
+                let tag = 1;
+                encode_key(tag, WireType::LengthDelimited, buf);
+                encode_varint(<Self as $crate::ReprBytes>::size() as u64, buf);
+                <Self as $crate::ReprBytes>::map_bytes(self, |bytes| buf.put_slice(bytes));
+            }
+
+            #[inline]
+            fn merge_field<B>(
+                &mut self,
+                tag: u32,
+                wire_type: ::prost::encoding::WireType,
+                buf: &mut B,
+                ctx: ::prost::encoding::DecodeContext,
+            ) -> Result<(), ::prost::DecodeError>
+            where
+                B: ::prost::bytes::Buf,
+            {
+                use ::alloc::string::ToString;
+                use ::prost::encoding::*;
+                use core::convert::TryInto;
+                if tag == 1 {
+                    let expected_size = <Self as $crate::ReprBytes>::size();
+
+                    check_wire_type(WireType::LengthDelimited, wire_type)?;
+                    let len = decode_varint(buf)?;
+                    if len > buf.remaining() as u64 {
+                        return Err(::prost::DecodeError::new("buffer underflow"));
+                    }
+                    if len != expected_size as u64 {
+                        return Err(::prost::DecodeError::new(
+                            LengthMismatch {
+                                expected: expected_size,
+                                found: len as usize,
+                            }
+                            .to_string(),
+                        ));
+                    }
+                    let result = <Self as $crate::ReprBytes>::from_bytes(
+                        (&buf.bytes()[0..expected_size]).try_into().unwrap(),
```

Comment:
What's try_into converting from/to here?

Reply:
Tries to copy the slice into an array of the size required by from_bytes in this ccontext. As the slice may have the wrong length, this is fallible. The unwrap should be an expect, tho.

Reply:
As noted, the `unwrap()` should be an `expect()`.